File: api/src/main/java/whenyourcar/api/apiPayload/ApiResponse.java
package whenyourcar.api.apiPayload;

import whenyourcar.common.code.BaseCode;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
@JsonPropertyOrder({"isSuccess", "whenyourcar/common/code", "message", "result"})
public class ApiResponse<T> {

    @JsonProperty("isSuccess")
    private final Boolean isSuccess;
    private final String code;
    private final String message;
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private T result;

    public static <T> ApiResponse<T> onSuccess(BaseCode code, T result){
        return new ApiResponse<>(true, code.getReasonHttpStatus().getCode() , code.getReasonHttpStatus().getMessage(), result);
    }

    public static <T> ApiResponse<T> onFailure(String code, String message, T data){
        return new ApiResponse<>(false, code, message, data);
    }
}

File: common/src/main/kotlin/howmuchcar/common/apiPayload/ApiResponse.kt
package howmuchcar.common.apiPayload

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import howmuchcar.common.code.status.SuccessStatus
import java.time.LocalDateTime

@JsonPropertyOrder("timestamp", "code", "message", "result")
data class ApiResponse<T> (
    @JsonProperty("timestamp")
    val timestamp:LocalDateTime = LocalDateTime.now(),
    val code:String,
    val message:String,

    @JsonProperty("result")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val result:T? = null
) {
    companion object {
        fun <T> onSuccess(result: T?): ApiResponse<T> {
            return ApiResponse(
                code = SuccessStatus._OK.getCode().code,
                message = SuccessStatus._OK.getCode().message,
                result = result
            )
        }

        fun <T> onFailure(
            code: String,
            message: String,
            data: T?
        ): ApiResponse<T> {
            return ApiResponse(
                code = code,
                message = message,
                result = data
            )
        }
    }
}